tests/pack/filelimit.c

```diff
@@ -0,0 +1,138 @@
+}
+
+void test_pack_filelimit__initialize_unlimited(void)
+{
+	_expected_open_mwindow_files = 15;
+	cl_git_pass(git_libgit2_opts(GIT_OPT_GET_MWINDOW_FILE_LIMIT, &_original_mwindow_file_limit));
+	cl_git_pass(git_libgit2_opts(GIT_OPT_SET_MWINDOW_FILE_LIMIT, 0));
+}
+
+void test_pack_filelimit__cleanup(void)
+{
+	git_buf path = GIT_BUF_INIT;
+	cl_git_pass(git_libgit2_opts(GIT_OPT_SET_MWINDOW_FILE_LIMIT, _original_mwindow_file_limit));
+
+	cl_git_pass(git_buf_joinpath(&path, clar_sandbox_path(), "repo.git"));
+	cl_fixture_cleanup(path.ptr);
+	git_buf_dispose(&path);
+}
+
+/*
+ * Create a packfile with one commit, one tree, and two blobs. The first blob
+ * (README.md) has the same content in all commits, but the second one
+ * (file.txt) has a different content in each commit.
+ */
+void create_packfile_commit(
+		git_repository *repo,
+		git_oid *out_commit_id,
+		git_oid *parent_id,
+		size_t commit_index,
+		size_t commit_count)
+{
+	git_buf file_contents = GIT_BUF_INIT;
+	git_treebuilder *treebuilder;
+	git_packbuilder *packbuilder;
+	git_signature *s;
+	git_oid oid, tree_id, commit_id;
+	const git_oid *parents[] = { parent_id };
+	size_t parent_count = parent_id ? 1 : 0;
+
+	cl_git_pass(git_treebuilder_new(&treebuilder, repo, NULL));
+
+	cl_git_pass(git_blob_create_from_buffer(&oid, repo, "", 0));
+	cl_git_pass(git_treebuilder_insert(NULL, treebuilder, "README.md", &oid, 0100644));
+
+	cl_git_pass(git_buf_printf(&file_contents, "Commit %zd/%zd", commit_index, commit_count));
+	cl_git_pass(git_blob_create_from_buffer(&oid, repo, file_contents.ptr, file_contents.size));
+	cl_git_pass(git_treebuilder_insert(NULL, treebuilder, "file.txt", &oid, 0100644));
+
+	cl_git_pass(git_treebuilder_write(&tree_id, treebuilder));
+	cl_git_pass(git_signature_now(&s, "alice", "alice@example.com"));
+	cl_git_pass(git_commit_create_from_ids(&commit_id, repo, "refs/heads/master", s, s,
+			NULL, file_contents.ptr, &tree_id, parent_count, parents));
+
+	cl_git_pass(git_packbuilder_new(&packbuilder, repo));
+	cl_git_pass(git_packbuilder_insert_commit(packbuilder, &commit_id));
+	cl_git_pass(git_packbuilder_write(packbuilder, NULL, 0, NULL, NULL));
+
+	cl_git_pass(git_oid_cpy(out_commit_id, &commit_id));
+
+	git_buf_dispose(&file_contents);
+	git_treebuilder_free(treebuilder);
+	git_packbuilder_free(packbuilder);
+	git_signature_free(s);
+}
+
+void test_pack_filelimit__open_repo_with_multiple_packfiles(void)
+{
+	git_buf path = GIT_BUF_INIT;
+	git_mwindow_ctl *ctl = &git_mwindow__mem_ctl;
+	git_repository *repo;
+	git_revwalk *walk;
+	git_oid id, *parent_id = NULL;
+	size_t i;
+	const size_t commit_count = 16;
+	unsigned int open_windows;
+
+	/*
+	 * Create a repository and populate it with 16 commits, each in its own
+	 * packfile.
+	 */
+	cl_git_pass(git_buf_joinpath(&path, clar_sandbox_path(), "repo.git"));
+	cl_git_pass(git_repository_init(&repo, path.ptr, true));
+	for (i = 1; i <= commit_count; ++i) {
```

Comment:
Micro-nit: we usually start iterating from `0`, not `1`. You can then just pass `i+1` to `create_packfile_commit`.